packages/next/src/client/components/router-reducer/reducers/navigate-reducer.ts

```diff
@@ -93,7 +93,211 @@ function addRefetchToLeafSegments(
 
   return appliedPatch
 }
-export function navigateReducer(
+
+// These implementations are expected to diverge significantly, so I've forked
+// the entire function. The one that's disabled should be dead code eliminated
+// because the check here is statically inlined at build time.
+export const navigateReducer = process.env.__NEXT_PPR
+  ? navigateReducer_PPR
+  : navigateReducer_noPPR
+
+// This is the implementation when PPR is disabled. We can assume its behavior
+// is relatively stable because it's been running in production for a while.
+function navigateReducer_noPPR(
+  state: ReadonlyReducerState,
+  action: NavigateAction
+): ReducerState {
+  const { url, isExternalUrl, navigateType, shouldScroll } = action
+  const mutable: Mutable = {}
+  const { hash } = url
+  const href = createHrefFromUrl(url)
+  const pendingPush = navigateType === 'push'
+  // we want to prune the prefetch cache on every navigation to avoid it growing too large
+  prunePrefetchCache(state.prefetchCache)
+
+  mutable.preserveCustomHistoryState = false
+
+  if (isExternalUrl) {
+    return handleExternalUrl(state, mutable, url.toString(), pendingPush)
+  }
+
+  let prefetchValues = state.prefetchCache.get(createHrefFromUrl(url, false))
+
+  // If we don't have a prefetch value, we need to create one
+  if (!prefetchValues) {
+    const data = fetchServerResponse(
+      url,
+      state.tree,
+      state.nextUrl,
+      state.buildId,
+      // in dev, there's never gonna be a prefetch entry so we want to prefetch here
+      // in order to simulate the behavior of the prefetch cache
+      process.env.NODE_ENV === 'development' ? PrefetchKind.AUTO : undefined
+    )
+
+    const newPrefetchValue = {
+      data,
+      // this will make sure that the entry will be discarded after 30s
+      kind:
+        process.env.NODE_ENV === 'development'
+          ? PrefetchKind.AUTO
+          : PrefetchKind.TEMPORARY,
+      prefetchTime: Date.now(),
+      treeAtTimeOfPrefetch: state.tree,
+      lastUsedTime: null,
+    }
+
+    state.prefetchCache.set(createHrefFromUrl(url, false), newPrefetchValue)
+    prefetchValues = newPrefetchValue
+  }
+
+  const prefetchEntryCacheStatus = getPrefetchEntryCacheStatus(prefetchValues)
+
+  // The one before last item is the router state tree patch
+  const { treeAtTimeOfPrefetch, data } = prefetchValues
+
+  prefetchQueue.bump(data!)
+
+  return data!.then(
+    ([flightData, canonicalUrlOverride, postponed]) => {
```

Comment:
It should be safe to remove the `postponed` handling in the non-PPR case, but I don't mind it being a separate PR if that's the plan 👍

Reply:
yeah I left it in the spirit of "do nothing except copy paste the entire implementation"